File: Fortran/StationaryDistribution.f90
SUBROUTINE StationaryDistribution

USE Parameters
USE Globals
USE Procedures
USE mUMFPACK

IMPLICIT NONE

INTEGER			:: ia,ib,iy,iaby,iab,it,iw(naby)
INTEGER, ALLOCATABLE	:: iwk(:)
REAL(8) 		:: ldiff
REAL(8), DIMENSION(nab) 	:: ldavec,ldbvec,ldiag,lgmat
REAL(8), DIMENSION(ngpa,ngpb) 	:: labdist
REAL(8), DIMENSION(ngpy,ngpy) 	:: leye,lmat
REAL(8), DIMENSION(nab,ngpy) 	:: lgmat1
TYPE(tCSR_di), DIMENSION(ngpy) 	 :: AUMF     !umfpack type

!identity matrix
leye = 0.0; DO iy = 1,ngpy
	leye(iy,iy) = 1.0
END DO


!$OMP PARALLEL DO
DO iab = 1,nab
	ldavec(iab) = adelta(afromab(iab))
	ldbvec(iab) = bdelta(bfromab(iab))
END DO
!$OMP END PARALLEL DO


IF	((EquilibriumR==0 .and. calibrating==.false.) &
  	.or. (EquilibriumR==1 .and. neqmiter<=100 .and. calibrating==.false.) &
	.or. (CalibrateDiscountRate==1 .and. neqmiter<=100 .and. calibrating==.false.) &
 	.or. (CalibrateRhoAtInitialGuess==1 .and. neqmiter<=100 .and. calibrating==.false.) &
	.or. (calibrating==.true. .and. ImposeEqumInCalibration==1  .and. neqmiter==1 ) &
	.or. (calibrating==.true. .and. ImposeEqumInCalibration==0 ) ) THEN

	gmat = 0.0

	!$OMP PARALLEL DO PRIVATE(iab,ia,ib)
	DO iy = 1,ngpy
		DO iab = 1,nab
			ia = afromab(iab)
			ib = bfromab(iab)
			IF(deathrate==0.0) THEN
				IF(Borrowing==0 .and. ia==1 .and. ib==2) gmat(iab,iy) = ydist(iy)
				IF(Borrowing==1 .and. ia==1 .and. ib==ngpbNEG+2) gmat(iab,iy) = ydist(iy)
!				IF(Borrowing==1 .and. ia==2 .and. ib==ngpbNEG+2) gmat(iab,:) = ydist
			ELSE IF(deathrate>0.0) THEN
				IF(Borrowing==0 .and. ia==1 .and. ib==1) gmat(iab,iy) = ydist(iy)
				IF(Borrowing==1 .and. ia==1 .and. ib==ngpbNEG+1) gmat(iab,iy) = ydist(iy)
			END IF

		END DO
 		gmat(:,iy) = ydist(iy)*gmat(:,iy)/SUM(gmat(:,iy)*ldavec*ldbvec)
	END DO
	!$OMP END PARALLEL DO
END IF

!$OMP PARALLEL DO PRIVATE(iwk,iw,ldiag)
DO iy = 1,ngpy
	IF(ALLOCATED(iwk)) DEALLOCATE(iwk)
	
	IF(ALLOCATED(BCSR(iy)%val)) DEALLOCATE(BCSR(iy)%val)
	IF(ALLOCATED(BCSR(iy)%col)) DEALLOCATE(BCSR(iy)%col)
	IF(ALLOCATED(BCSR(iy)%row)) DEALLOCATE(BCSR(iy)%row)
	
	ALLOCATE(iwk(AUCSR(iy)%nz))
	ALLOCATE(BCSR(iy)%val(AUCSR(iy)%nz), BCSR(iy)%col(AUCSR(iy)%nz), BCSR(iy)%row(AUCSR(iy)%n+1))
	
	!transpose AU matrix
	BCSR(iy) = AUCSR(iy)
	CALL csrcsc (nab, 1, 1, AUCSR(iy)%val, AUCSR(iy)%col, AUCSR(iy)%row, BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row )

	!adjust A' matrix for non-linearly spaced grids
	ldiag = 1.0/(ldavec*ldbvec)
	CALL diamua (nab, 1, BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row, ldiag,BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row )
	ldiag = ldavec*ldbvec
	CALL amudia (nab, 1, BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row, ldiag,BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row )
	

 	BCSR(iy)%val = -deltakfe*BCSR(iy)%val

	ldiag = 1.0  - deltakfe*ymarkovdiag(iy,iy) + deltakfe*deathrate
	CALL apldia (nab, 0, BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row, ldiag, BCSR(iy)%val, BCSR(iy)%col, BCSR(iy)%row, iw )
	
	AUMF(iy) = tCSR_di(BCSR(iy)%row-1,BCSR(iy)%col-1,BCSR(iy)%val) 
	
	
END DO
!$OMP END PARALLEL DO


lmat = leye + deltakfe*TRANSPOSE(ymarkovoff) 

ldiff = 1.0
it = 1

DO WHILE (ldiff>KFEtol .and. it<maxiterKFE)

	!sweep over y
	!$OMP PARALLEL DO PRIVATE(lgmat)
	DO iy = 1,ngpy	

 		lgmat = MATMUL(gmat(:,:),lmat(iy,:))
 		lgmat = lgmat + deltakfe*deathrate*MERGE(SUM(gmat(:,iy)*ldavec*ldbvec),0.0_8,afromab==1 .and. bfromab==ngpbNEG+1 )/(ldavec*ldbvec)
		lgmat1(:,iy) = AUMF(iy) .umfpack. lgmat		
		
	END DO
	!$OMP END PARALLEL DO
		
	ldiff = maxval(abs(gmat-lgmat1))
	gmat = lgmat1
	it = it+1
	IF (Display>=2) write(*,*) 'KFE iteration ',it,' max g change: ',ldiff
END DO

gmat = MERGE(0.0_8,gmat,abs(gmat)<1.0e-50_8)

!$OMP PARALLEL DO PRIVATE(ia,ib,iy,iab)	
DO iaby = 1,naby
	ia = afromaby(iaby)
	ib = bfromaby(iaby)
	iy = yfromaby(iaby)
	iab = 	abfromab(ia,ib)
	gjoint(ia,ib,iy) = gmat(iab,iy)
	gvec(iaby) = gmat(iab,iy)
END DO
!$OMP END PARALLEL DO

!marginal distributions
DO ia = 1,ngpa
DO iy = 1,ngpy
	gamarg(ia,iy) = SUM(gjoint(ia,:,iy)*bdelta)
END DO
END DO

DO ib = 1,ngpb
DO iy = 1,ngpy
	gbmarg(ib,iy) = SUM(gjoint(:,ib,iy)*adelta)
END DO
END DO

!ab joint distribution
DO ib = 1,ngpb
	gabmarg(:,ib) = SUM(gjoint(:,ib,:),DIM=2)*adelta*bdelta(ib)
END DO

!ab cumulative distribution
labdist(1,:) = gabmarg(1,:)
DO ia =2,ngpa
	labdist(ia,:) = labdist(ia-1,:)+ gabmarg(ia,:)
END DO
gabcum(:,1) = labdist(:,1)
DO ib = 2,ngpb
	gabcum(:,ib) = gabcum(:,ib-1)+ labdist(:,ib)
END DO


END SUBROUTINE StationaryDistribution

File: mycode/StationaryDistribution.m
if (~EquilibriumR && ~calibrating) || ...
    (EquilibriumR && neqmiter<=100 && ~calibrating) || ...
    (CalibrateDiscountRate && neqmiter<=100 && ~calibrating) || ...
    (CalibrateRhoAtInitialGuess && neqmiter<=100 && ~calibrating) || ...
    (calibrating && ImposeEqumInCalibration && neqmiter==1 ) || ...
    (calibrating && ~ImposeEqumInCalibration)

    gmat = zeros(ngpa,ngpb,ngpy);
    ib = 1+~deathrate+Borrowing*ngpbNEG;
    gmat(1,ib,:) = ydist/abdelta(1,ib);
    gmat = reshape(gmat,nab,ngpy);
end

% a drift, upwind
laudriftB = min(d + adrift,0);
laudriftB(:,ngpb,:) = laudriftB(:,ngpb-1,:);
laudriftF = max(d + adrift,0);
laudriftF(:,ngpb,:) = laudriftF(:,ngpb-1,:);

% b drift,upwind
bdot = s - dadj;
lbudriftB = min(bdot,0);
lbudriftB(:,ngpb,:) = min(s(:,ngpb,:) - dadj(:,ngpb-1,:),0);
lbudriftF = max(bdot,0);
% lbudriftF(:,ngpb,:) = max(s(:,ngpb,:) - dadj(:,ngpb-1,:),0);

AU = reshape([ laudriftF./[dagrid;1] ... % a+1
              -laudriftB./[1;dagrid] ... % a-1
               lbudriftF./[dbgrid 1] ... % b+1
              -lbudriftB./[1 dbgrid] ... % b-1
            ], nab,4,ngpy);
B = [1+deltakfe*(sum(AU,2)+deathrate-ymarkovdiag) ... % diagonal value
      -deltakfe*AU.*ABdelta]; % adjust for non-linearly spaced grids

lmat = eye(ngpy) + deltakfe*ymarkovoff';

ldiff = 1.0;
it = 1;
ib = 1+ngpa*ngpbNEG;
while ldiff>KFEtol && it<maxiterKFE
    lgmat = gmat*lmat';
    lgmat(ib,:) = lgmat(ib,:) + deltakfe*deathrate/abdelta(ib)*(abdelta(:)'*gmat);
    % sweep over y
    for iy = 1:ngpy %par
        lgmat(:,iy) = spdiags(B(:,:,iy),[0 -1 1 -ngpa ngpa],nab,nab)\lgmat(:,iy);
    end
   
	ldiff = max(abs(gmat(:)-lgmat(:)));
	gmat = lgmat;
	it = it+1;
	if Display>1; fprintf('KFE iteration: %3d    max g change: %.15g\n',it,ldiff); end
end
gmat(abs(gmat)<1e-50) = 0;
gmat = reshape(gmat,ngpa,ngpb,ngpy);